torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -719,6 +719,50 @@ std::vector<Tensor> cat_tensors_backward(const Tensor & grad, const std::vector<
   return grad_inputs;
 }
 
+std::vector<Tensor> block_diag_backward(const Tensor & grad, const std::vector<std::vector<int64_t>> &sizes, const std::vector<ScalarType> &dtypes) {
+  std::vector<Tensor> grad_inputs(sizes.size());
+  if (!grad.defined()) {
+    return grad_inputs;
+  }
+  Tensor grad_;
+  bool grad_is_complex = grad.is_complex();
+  if (grad_is_complex) {
+    grad_ = at::real(grad);
+  }
```

Comment:
What is going on with complex numbers? Is the problem that there is a weird case when we have mixed dtype inputs to block_diag, some of which can be complex, and we need to handle that?

Reply:
yeah this is pretty interesting - I got this from the logic for `cat` backward, and it looks this handles exactly that - mixed complex/non-complex inputs. I tried a mixed complex/non-complex input locally and it only started passing when I added this.

Reply:
(there also aren't OpInfo tests for the mixed case, so I'll add some)

Reply:
That makes sense, I will read the formula more closely. I am not sure OpInfos are supposed to test the mixed dtype case (@mruberry ?) because the framework requests a test with a specific dtype. Maybe we just want a manual test somewhere for this.

Reply:
Oh I have this implemented by having the sample inputs function check if the passed-in dtype is complex, and if so it spits out an extra sample input for the [only complex], and [mixed complex/noncomplex] case, and the OpInfo suite seemed to pass locally for me. I'll see what happens in CI